### PR TITLE
feat(examples): hall of fame

### PR DESCRIPTION
**Pull request by leohhhn**
## Description

Depends on #2584 for `avlpager`

Introduces the `r/demo/hof` realm.

The Hall of Fame is an exhibition that holds items. Users can add their realms to the Hall of Fame by importing the Hall of Fame realm and calling `hof.Register()` from their `init` function. 

The realm is moderated and the registrations be paused at will.

Screenshot 2024-10-07 at 20 09 43

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
-  [x] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests
- [x] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Comment by leohhhn** [comment]
@moul 

Care for a prelim review before I go on?

File: examples/gno.land/r/gnoland/hof/gno.mod
Comment (moul) [inline examples/gno.land/r/gnoland/hof/gno.mod:1]: It shouldn't be "gnoland"; it's not "official" or "useful" enough, and it allows too much permissionless activity. 

Let's make it a demo.

File: examples/gno.land/r/gnoland/hof/hof.gno
Comment (moul) [inline examples/gno.land/r/gnoland/hof/hof.gno:14]: The admin should be able to pause the contract completely and also have the ability to delete some entries. Essentially, we allow it to run until it becomes spammed, at which point we clean and pause it. Later, we can revisit the hof contract to make it DAO-managed.

File: examples/gno.land/r/gnoland/hof/hof.gno
Comment (moul) [inline examples/gno.land/r/gnoland/hof/hof.gno:18]: i've a doubt, shoudl `avl.Tree` be a pointer?

File: examples/gno.land/r/gnoland/hof/hof.gno
Comment (moul) [inline examples/gno.land/r/gnoland/hof/hof.gno:55]: Maybe you can turn this multi-expo hall into a second demo while keeping the permanent collection as a straightforward mono-exhibit contract. I'm currently looking for demos that are as simple as possible.

File: examples/gno.land/r/gnoland/hof/hof.gno
Comment (thehowl) [reply examples/gno.land/r/gnoland/hof/hof.gno:18]: can be both as the methods use a pointer receiver, which can be "implicit". (same as declaring `var buf bytes.Buffer` and then using `buf.Write`).

I guess the advantage to not making it a pointer is that the zero value is ready to use.

File: examples/gno.land/r/gnoland/hof/hof.gno
Comment (leohhhn) [reply examples/gno.land/r/gnoland/hof/hof.gno:14]: good thing we have p/demo/pausable :)

File: examples/gno.land/r/gnoland/hof/hof.gno
Comment (leohhhn) [reply examples/gno.land/r/gnoland/hof/hof.gno:18]: `NewTree` returns a pointer; we should decide there first.

File: examples/gno.land/r/gnoland/hof/rendering.gno
Comment (moul) [inline examples/gno.land/r/gnoland/hof/rendering.gno:49]: should use a pager :)

File: examples/gno.land/r/gnoland/hof/hof.gno
Comment (moul) [inline examples/gno.land/r/gnoland/hof/hof.gno:78]:
```suggestion
func Register() {
```

what about making this contract a super simple one?

File: examples/gno.land/r/leon/home/home.gno
Comment (moul) [inline examples/gno.land/r/leon/home/home.gno:37]: multiple exhibitions; plenty returned values that are never used

please, try to make minimalist contracts; that's what we need at the moment

File: examples/gno.land/r/gnoland/hof/rendering.gno
Comment (moul) [inline examples/gno.land/r/gnoland/hof/rendering.gno:90]: i think this check should be removed

File: examples/gno.land/r/gnoland/hof/rendering.gno
Comment (moul) [inline examples/gno.land/r/gnoland/hof/rendering.gno:94]: i think this check should just return ""

File: examples/gno.land/r/gnoland/hof/rendering.gno
Comment (moul) [inline examples/gno.land/r/gnoland/hof/rendering.gno:89]:
```suggestion
func RenderExhibWidget(exhibID seqid.ID, itemsToRender int) string {
```

File: examples/gno.land/r/gnoland/home/home.gno
Comment (moul) [inline examples/gno.land/r/gnoland/home/home.gno:100]: Haha, my UI library is definitely very poor in terms of usability.

Comment (moul) [inline examples/gno.land/r/gnoland/home/home.gno:100]: Haha, my UI library is definitely very poor in terms of usability.

File: examples/gno.land/r/gnoland/hof/hof.gno
Comment (leohhhn) [reply examples/gno.land/r/gnoland/hof/hof.gno]: [19f7fa9](https://github.com/gnolang/gno/pull/2842/commits/19f7fa93995e3ca2352977708df3c38d2252bf7f)

File: examples/gno.land/r/gnoland/hof/hof.gno
Comment (leohhhn) [reply examples/gno.land/r/gnoland/hof/hof.gno:13]: [a646518](https://github.com/gnolang/gno/pull/2842/commits/a64651897956d45980086aa2f9cd054f0a50d6ea)

File: examples/gno.land/r/gnoland/hof/rendering.gno
Comment (leohhhn) [reply examples/gno.land/r/gnoland/hof/rendering.gno:31]: Let's add it after we merge your avl/pager PR

File: examples/gno.land/r/gnoland/hof/hof.gno
Comment (leohhhn) [reply examples/gno.land/r/gnoland/hof/hof.gno]: Removed the return values

File: examples/gno.land/r/leon/home/home.gno
Comment (leohhhn) [reply examples/gno.land/r/leon/home/home.gno:37]: Removed temporary exhibitions, just have one now.

File: examples/gno.land/r/gnoland/hof/rendering.gno
Comment (leohhhn) [reply examples/gno.land/r/gnoland/hof/rendering.gno:69]: [94fc361](https://github.com/gnolang/gno/pull/2842/commits/94fc3614551e167a1cd2a25d15d87388e13f1c56)

File: examples/gno.land/r/gnoland/hof/rendering.gno
Comment (leohhhn) [reply examples/gno.land/r/gnoland/hof/rendering.gno]: rm'd it

File: examples/gno.land/r/gnoland/hof/gno.mod
Comment (leohhhn) [reply examples/gno.land/r/gnoland/hof/gno.mod]: Moved it to `gno.land/r/demo/hof`

File: examples/gno.land/r/demo/hof/rendering.gno
Comment (moul) [inline examples/gno.land/r/demo/hof/rendering.gno:10]: for this, you'll be able to use #2887

File: examples/gno.land/r/demo/hof/hof.gno
Comment (moul) [inline examples/gno.land/r/demo/hof/hof.gno:61]: I think we should not panic about registration but instead return an error. 

If someone has a conditional registration that can occur 0 or N times, it will brick their contract. 

Additionally, when we add "upgradeable contracts," the init() function can be called multiple times.

I suggest returning an error but ignoring it from the caller by default, as we don't need to address it.

Comment (leohhhn) [reply examples/gno.land/r/demo/hof/hof.gno:61]: https://github.com/gnolang/gno/pull/2842#discussion_r1784961023

What about this? :D

File: examples/gno.land/p/demo/ownable/ownable.gno
Comment (leohhhn) [reply examples/gno.land/p/demo/ownable/ownable.gno:61]: Yep, missed it. Fixed: [f32070f](https://github.com/gnolang/gno/pull/2842/commits/f32070fe3b36ea1538c050060b4882630d810121)

File: examples/gno.land/r/gnoland/hof/rendering.gno
Comment (leohhhn) [reply examples/gno.land/r/gnoland/hof/rendering.gno]: Ok, will do.

Comment (leohhhn) [reply examples/gno.land/r/gnoland/hof/rendering.gno]: Using the pager now. Since it doesn't fully work yet, I've set my page size so that pages don't appear until we fix the pager.

File: examples/gno.land/r/demo/hof/hof.gno
Comment (moul) [reply examples/gno.land/r/demo/hof/hof.gno:60]: I'm sorry if I initially said something wrong; I don't recall the context. 

However, `Register` should not panic. Let's avoid returning anything and don't panic. 

The goal is to ensure you're registered at launch without needing complex logic to check if you were already registered. It's similar to a "Set" or "Upsert" in other systems (can override with the same value), or sending a "wakeup" UDP packet where it doesn't matter if the system is already up and running.

Comment (leohhhn) [reply examples/gno.land/r/demo/hof/hof.gno]: [8bcf672](https://github.com/gnolang/gno/pull/2842/commits/8bcf672fe64a26fe6a6ad307bf04a732282b3504)

File: examples/gno.land/p/demo/fqname/fqname_test.gno
Comment (ajnavarro) [inline examples/gno.land/p/demo/fqname/fqname_test.gno:86]: `gno.land/p/demo/avl?foo=test1&bar=test2`?

Comment (ajnavarro) [reply examples/gno.land/p/demo/fqname/fqname_test.gno:86]: `/p/demo/avl`?

File: examples/gno.land/p/demo/ownable/ownable.gno
Comment (ajnavarro) [inline examples/gno.land/p/demo/ownable/ownable.gno:40]: Is not the same as before?

File: examples/gno.land/r/demo/hof/hof.gno
Comment (ajnavarro) [inline examples/gno.land/r/demo/hof/hof.gno:94]: why not just `true` instead of `struct{}{}`? If you want to be picky about using less space you should use `interface{}`

File: examples/gno.land/r/demo/hof/hof.gno
Comment (ajnavarro) [inline examples/gno.land/r/demo/hof/hof.gno:97]: Can you downvote and upvote the same item?

File: examples/gno.land/r/demo/hof/hof.gno
Comment (ajnavarro) [inline examples/gno.land/r/demo/hof/hof.gno:110]: If you can only upvote or downvote, why don't just `item.vote.Set(caller, UP)`?

File: examples/gno.land/r/demo/hof/hof_test.gno
Comment (ajnavarro) [inline examples/gno.land/r/demo/hof/hof_test.gno:98]:
```suggestion
	// Check double downvote
```

File: examples/gno.land/r/demo/hof/hof_test.gno
Comment (ajnavarro) [inline examples/gno.land/r/demo/hof/hof_test.gno:77]: Check this test, it is badly copy/pasted from the upvote one.

File: examples/gno.land/r/demo/hof/rendering.gno
Comment (ajnavarro) [inline examples/gno.land/r/demo/hof/rendering.gno:103]: this is not needed. If there are no items, the output will be "" anyway.

File: examples/gno.land/p/demo/ownable/ownable.gno
Comment (leohhhn) [reply examples/gno.land/p/demo/ownable/ownable.gno:40]: yes, just thought its better to use the methods available on `std.Address`

File: examples/gno.land/r/demo/hof/hof.gno
Comment (leohhhn) [reply examples/gno.land/r/demo/hof/hof.gno:94]: Isn't `struct{}{}` the most optimal memory-wise (0 bytes)?

File: examples/gno.land/r/demo/hof/hof.gno
Comment (leohhhn) [reply examples/gno.land/r/demo/hof/hof.gno:97]: Yes, it's okay

File: examples/gno.land/r/demo/hof/hof.gno
Comment (leohhhn) [reply examples/gno.land/r/demo/hof/hof.gno:110]: It's easier to get the num of upvotes/downvotes if we have two trees and just get their size; otherwise you'd have to iterate and count

File: examples/gno.land/r/demo/hof/rendering.gno
Comment (leohhhn) [reply examples/gno.land/r/demo/hof/rendering.gno]: You're right; rm'd the condition: [ff06493](https://github.com/gnolang/gno/pull/2842/commits/ff06493e12c7dae579ea2c0cbcb9d3eee276f4f7)

File: examples/gno.land/r/demo/hof/hof_test.gno
Comment (leohhhn) [reply examples/gno.land/r/demo/hof/hof_test.gno:77]: I checked it; only comments were wrong; otherwise it should be good, no? [ff06493](https://github.com/gnolang/gno/pull/2842/commits/ff06493e12c7dae579ea2c0cbcb9d3eee276f4f7)

File: examples/gno.land/p/demo/fqname/fqname_test.gno
Comment (leohhhn) [reply examples/gno.land/p/demo/fqname/fqname_test.gno:86]: Yeah. This whole package needs to be tested better; I'll add a few cases and more docs about the func

Comment (leohhhn) [reply examples/gno.land/p/demo/fqname/fqname_test.gno]: rm'd my addition to fqname; i just needed this simple happy-path flow, which is happening 100% of the time in my case. [c685a46](https://github.com/gnolang/gno/pull/2842/commits/c685a469359c6c29173a194f9f1e88d3b3d1ae08)

**Review by moul — APPROVED** [review]

File: examples/gno.land/r/demo/hof/hof.gno
Comment (ajnavarro) [reply examples/gno.land/r/demo/hof/hof.gno:96]: I don't really know if that is true on GnoVM

File: examples/gno.land/r/demo/hof/hof.gno
Comment (ajnavarro) [reply examples/gno.land/r/demo/hof/hof.gno:99]: If that is the expected behavior it is good for me.

File: examples/gno.land/r/demo/hof/hof.gno
Comment (leohhhn) [reply examples/gno.land/r/demo/hof/hof.gno:96]: How can we figure it out?

**Pull request merged**
